plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+        E = np.array(
+            [Ex.to(u.V / u.m).value, Ey.to(u.V / u.m).value, Ez.to(u.V / u.m).value]
+        )
+        E = np.moveaxis(E, 0, -1)
+        B = np.array([Bx.to(u.T).value, By.to(u.T).value, Bz.to(u.T).value])
+        B = np.moveaxis(B, 0, -1)
+
+        # Calculate the adaptive time step from the fields currently experienced
+        # by the particles
+        # If user sets dt explicitly, that's handled in _adaptive_dt
+        dt = self._adaptive_dt(Ex, Ey, Ez, Bx, By, Bz)
+
+        # TODO: Test v/c and implement relativistic Boris push when required
+        # vc = np.max(v)/_c
+
+        # Make sure the time step can be multiplied by a [nparticles, 3] shape field array
+        if dt.size > 1:
+            dt = dt[tracked_mask, np.newaxis]
+
+            # Increment the tracked particles' time by dt
+            self.time[tracked_mask] += dt
+        else:
+            self.time += dt
+
+        self.x[tracked_mask], self.v[tracked_mask] = boris_push(
+            pos_tracked, vel_tracked, B, E, self.q, self.m, dt, inplace=False
+        )
+
+        self.dt = dt
+
+    @property
+    def on_any_grid(self):
+        """
+        Binary array for each particle indicating whether it is currently
+        on ANY grid.
+        """
+        return np.sum(self.on_grid, axis=-1) > 0
+
+    @property
+    def vmax(self):
+        """
+        Calculate the maximum velocity.
+        Used for determining the grid crossing maximum time step.
+        """
+        tracked_mask = self._tracked_particle_mask()
+
+        return np.max(np.linalg.norm(self.v[tracked_mask], axis=-1))
+
+    def _validate_constructor_inputs(
+        self, termination_condition, save_routine, dt, dt_range, field_weighting: str
+    ):
+        """
+        Ensure the specified termination condition and save routine are actually
+        a termination routine class and save routine, respectively. This function also
+        sets the `_is_synchronized_time_step` and `_is_adaptive_time_step` attributes.
+        """
+        if not isinstance(termination_condition, AbstractTerminationCondition):
+            raise TypeError("Please specify a valid termination condition.")
+
+        if save_routine is not None and not isinstance(
+            save_routine, AbstractSaveRoutine
+        ):
+            raise TypeError("Please specify a valid save routine")
+
+        require_synchronized_time = termination_condition.require_synchronized_dt or (
+            save_routine is not None and save_routine.require_synchronized_dt
+        )
+
+        if dt is not None and dt_range is not None:
+            raise ValueError("Please only specify either dt or dt_range")
+
+        # Will the simulation follow a synchronized time step?
+        # This must be the case for certain termination conditions and saving routines
+        if isinstance(dt, u.Quantity):
+            if isinstance(dt.value, np.ndarray):
```

Comment:
This logic looks like it needs to be updated to use the new `dt_range` variable? 

An exception should be raised here if you try to set `dt` AND `dt_range`